drivers/pppou.c

```diff
@@ -619,152 +479,167 @@ static int pppos_netifUp(pppos_priv_t *state)
+			*arg++ = '\0';
+			break;
+		}
+	}
 
-	// NOTE: netif->state cannot be used to keep our private state as it is used by LWiP PPP implementation, pass it as *ctx to callbacks
-	state = netif->state;
-	netif->state = NULL;
+	return arg;
+}
 
-	memset(state, 0, sizeof(pppos_priv_t));
-	state->netif = netif;
-	state->serialdev_fn = cfg;
 
-	while (*cfg && *cfg != ':')
-		++cfg;
+static int pppou_netifInit(struct netif *netif, char *cfg)
+{
+	char *next;
+	pppou_priv_t *state;
+	int retries, flags = 0;
 
-	if (*cfg != ':') {
-		log_error("APN is not configured");
-		return ERR_ARG;
+	if (cfg == NULL || *cfg == '\0') {
+		log_error("no config");
+		return ERR_IF;
 	}
 
-	*cfg = 0;
-	cfg++;
+	/*
+	 * NOTE: netif->state cannot be used to keep our private state as it is used
+	 * by LWiP PPP implementation, pass it as *ctx to callbacks
+	 */
+	state = netif->state;
+	netif->state = NULL;
 
-#if PPPOS_USE_CONFIG_FILE
-	state->config_path = cfg;
-#endif
+	memset(state, 0, sizeof(pppou_priv_t));
 
-	while (*cfg && *cfg != ':')
-		++cfg;
+	state->serial_speed = B115200;
+	state->netif = netif;
+	state->fd = -1;
 
-	if (*cfg == ':') {
-		*cfg = 0;
-		cfg++;
-		state->serialat_fn = cfg;
-	}
-	else {
-		state->serialat_fn = "/dev/ttyacm1";
+	for (; (next = cfg_get_next_arg(cfg)); cfg = next) {
+		speed_t speed = serial_speed_from_string(cfg);
+		if (speed > 0) {
+			state->serial_speed = speed;
+			log_info("config speed: %s bps => %d", cfg, speed);
+			continue;
+		}
+
+		if (!strncmp(cfg, "/dev/", 5)) {
+			state->serial_dev = cfg;
+			log_info("config device: ", cfg);
+			continue;
+		}
+
+		if (strcmp(cfg, "up") == 0) {
+			flags |= CFG_FLAG_DEFAULT_UP;
+			log_info("config up: yes");
+			continue;
+		}
+
+		/* TODO: extend with other flags */
 	}
 
-	state->fd = -1;
+	if (!state->serial_dev) {
+		log_error("no device");
+		return ERR_IF;
+	}
 
 	mutexCreate(&state->lock);
 	condCreate(&state->cond);
 
+	/* FIXME: resolve `pp` name conflict with `pppos`
+	 * and is required for e.g. netif_is_ppp() */
 	netif->name[0] = 'p';
 	netif->name[1] = 'p';
 
-	if (!cfg)
-		return ERR_ARG;
-
 	if (!state->ppp) {
-		state->ppp = pppapi_pppos_create(state->netif, pppos_output_cb, pppos_link_status_cb, state);
+		state->ppp = pppapi_pppos_create(state->netif, pppou_output_cb, pppou_link_status_cb, state);
 
 		if (!state->ppp) {
 			log_error("could not create PPP control interface");
-			return ERR_IF ; // TODO: maybe permanent broken state?
+			return ERR_IF; /* TODO: maybe permanent broken state? */
 		}
 		netif->flags &= ~NETIF_FLAG_UP;
-		ppp_set_netif_statuscallback(state->ppp, pppos_statusCallback);
+		ppp_set_netif_statuscallback(state->ppp, pppou_statusCallback);
 	}
 
-	beginthread(pppos_mainLoop, 4, (void *)state->main_loop_stack, sizeof(state->main_loop_stack), state);
-
-	return ERR_OK;
-}
+	beginthread(pppou_mainLoop, 4, (void *)state->main_loop_stack, sizeof(state->main_loop_stack), state);
 
+	for (retries = 3; retries > 0; retries--) {
```

Comment:
I assume this is temporary because of lack of real `ifconfig` on the target platform (usually I would not like for netif initialization to be unnecessarily blocking)

Reply:
That's right.